src/main/java/io/temporal/internal/statemachines/ChildWorkflowStateMachine.java

```diff
@@ -0,0 +1,308 @@
+
+  /**
+   * Cancellation through this class is valid only when start child workflow command is not sent
+   * yet. Cancellation of an initiated child workflow is done through CancelExternalCommands. So all
+   * of the types besides ABANDON are treated differently.
+   */
+  public void cancel() {
+    if (!isFinalState()) {
+      explicitEvent(ExplicitEvent.CANCEL);
+    }
+  }
+
+  private void cancelStartChildCommand() {
+    cancelCommand();
+    RuntimeException failure =
+        new ChildWorkflowFailure(
+            0,
+            0,
+            startAttributes.getWorkflowType().getName(),
+            WorkflowExecution.newBuilder().setWorkflowId(startAttributes.getWorkflowId()).build(),
+            startAttributes.getNamespace(),
+            RetryState.RETRY_STATE_NON_RETRYABLE_FAILURE,
+            new CanceledFailure("Child immediately canceled", null, null));
+    completionCallback.apply(Optional.empty(), failure);
+  }
+
+  private void notifyCompleted() {
+    ChildWorkflowExecutionCompletedEventAttributes attributes =
+        currentEvent.getChildWorkflowExecutionCompletedEventAttributes();
+    Optional<Payloads> result =
+        attributes.hasResult() ? Optional.of(attributes.getResult()) : Optional.empty();
+    completionCallback.apply(result, null);
+  }
+
+  private void notifyStartFailed() {
+    StartChildWorkflowExecutionFailedEventAttributes attributes =
+        currentEvent.getStartChildWorkflowExecutionFailedEventAttributes();
+    Exception failure =
+        new ChildWorkflowTaskFailedException(
+            currentEvent.getEventId(),
+            WorkflowExecution.newBuilder().setWorkflowId(attributes.getWorkflowId()).build(),
+            attributes.getWorkflowType(),
+            RetryState.RETRY_STATE_NON_RETRYABLE_FAILURE,
+            null);
+    failure.initCause(
+        new WorkflowExecutionAlreadyStarted(
+            WorkflowExecution.newBuilder().setWorkflowId(attributes.getWorkflowId()).build(),
+            attributes.getWorkflowType().getName(),
+            null));
+    completionCallback.apply(Optional.empty(), failure);
+  }
+
+  private void notifyFailed() {
+    ChildWorkflowExecutionFailedEventAttributes attributes =
+        currentEvent.getChildWorkflowExecutionFailedEventAttributes();
+    RuntimeException failure =
+        new ChildWorkflowTaskFailedException(
+            currentEvent.getEventId(),
+            attributes.getWorkflowExecution(),
+            attributes.getWorkflowType(),
+            attributes.getRetryState(),
+            attributes.getFailure());
+    completionCallback.apply(Optional.empty(), failure);
+  }
+
+  private void notifyTimedOut() {
+    ChildWorkflowExecutionTimedOutEventAttributes attributes =
+        currentEvent.getChildWorkflowExecutionTimedOutEventAttributes();
+    TimeoutFailure timeoutFailure =
+        new TimeoutFailure(null, null, TimeoutType.TIMEOUT_TYPE_START_TO_CLOSE);
+    timeoutFailure.setStackTrace(new StackTraceElement[0]);
```

Comment:
why do we need to set the stack trace here, but not for terminated?

Reply:
removed